File: Cpp/LeetCode202.HappyNumber.cpp
#include <unordered_set>

class Solution {
public:
    bool isHappy(int n) {
        std::unordered_set <int> inSet;
        while(true)
        {
            if (n==1) {
                return true;
            }
            int sum = 0;
            int mod = 0;
            while(n>0)
            {
                mod = n%10;
                n = n/10;
                sum += mod*mod;
            }
            if (inSet.find(sum) != inSet.end())
            {
                return false;
            }
            inSet.insert(sum);
            n = sum;
        }
        return false;
    }
};

File: Python/LeetCode202.HappyNumber.py
class Solution:
    def getSquare(self, n:int):
        m = 0
        while n > 0:
            x = n % 10
            n = n // 10
            m = m + (x*x)
        return m
    def isHappy(self, n: int) -> bool:
        slow = self.getSquare(n)
        fast = self.getSquare(n)
        fast = self.getSquare(fast)
        while (slow != fast):
            slow = self.getSquare(slow)
            fast = self.getSquare(fast)
            fast = self.getSquare(fast)
        if (slow == 1):
            return True
        else:
            return False
